### ISSUE TITLE
Question about the performance of CT-RATE

### ISSUE THREAD

**Issue by laihaoran** [alibaba-damo-academy/fvlm#12]
Thanks for your code and weights. We strictly followed your released code and loaded your publicly available weights to perform inference on the CT-RATE dataset. While some disease categories achieve AUC results that are consistent and perform well, many lung-related diseases show much lower AUC values in our experiments.

Our obtained results are as follows (valid samples: 3039 for each category):

Emphysema                          : 0.4954
Atelectasis                        : 0.3954
Lung nodule                        : 0.5535
Lung opacity                       : 0.5840
Pulmonary fibrotic sequela         : 0.4885
Pleural effusion                   : 0.4820
Mosaic attenuation pattern         : 0.7851
Peribronchial thickening           : 0.4906
Consolidation                      : 0.6232
Bronchiectasis                     : 0.5797
Interlobular septal thickening     : 0.7462
Cardiomegaly                       : 0.8978
Pericardial effusion               : 0.7417
Coronary artery wall calcification : 0.7480
Hiatal hernia                      : 0.6392
Arterial wall calcification        : 0.7771

Macro AUC (ignoring NA): 0.6267  
Micro AUC: 0.5772  
Could you please share the actual AUC results you obtained for these categories, and provide any possible reasons why the lung-related disease performance might be significantly lower compared to others?

Thank you for your time and support.

**Comment by fepegar** [comment]
@laihaoran would you be able to share your code? I'm struggling to use this repo:
- https://github.com/alibaba-damo-academy/fvlm/issues/11

**Comment by lavsendahal** [comment]
I haven’t yet looked at the AUC results at the disease level, but I did reproduce the average results reported for the CT-RATE dataset. I haven’t run validation on the RAD-ChestCT dataset yet.

@fepegar I can share my code if it’s not working on your side. Just a heads-up: there are still a lot of hard-coded paths I haven’t cleaned up, and some library mismatches, so it’ll take a bit of work to make it more consistent.

**Comment by fepegar** [comment]
Thanks @lavsendahal, that'd be great!

**Comment by lavsendahal** [comment]
@fepegar  Here’s my forked copy with the changes I made:

https://github.com/lavsendahal/fvlm/tree/main

**_Dataset handling_**

Downloaded Totalsegmentator masks from Hugging Face instead of computing them locally (faster).

Downloaded train_fixed and valid_fixed splits from Hugging Face (already corrected), so I skipped running fix_data.py.

Only ran preprocess.py and resize.py from the data/ directory before inference.

**_Environment / dependencies_**

Used Python 3.10 as the base environment (compatibility reasons).

Modified requirements.txt to resolve library conflicts.

**_Model configuration_**

_Edited the BERT JSON config and added:

"add_type_embeddings": true_


**Other notes**

Code paths are still hard-coded, as in the original authors’ version.
Also need folder structure as suggested in original repo. 
This is work in progress, but you can check my fork and let me know if you run into any issues.
Need microsoft/BiomedVLP-CXR-BERT-specialized from huggingface at code root folder right now. 
The results of my run in CT_Rate validation set are in the folder rate_res/. 
Can run compute_metrics.py to get AUC and other performance metrics.

**Comment by jz5426** [comment]
> [@fepegar](https://github.com/fepegar) Here’s my forked copy with the changes I made:
> 
> https://github.com/lavsendahal/fvlm/tree/main
> 
> **_Dataset handling_**
> 
> Downloaded Totalsegmentator masks from Hugging Face instead of computing them locally (faster).
> 
> Downloaded train_fixed and valid_fixed splits from Hugging Face (already corrected), so I skipped running fix_data.py.
> 
> Only ran preprocess.py and resize.py from the data/ directory before inference.
> 
> **_Environment / dependencies_**
> 
> Used Python 3.10 as the base environment (compatibility reasons).
> 
> Modified requirements.txt to resolve library conflicts.
> 
> **_Model configuration_**
> 
> _Edited the BERT JSON config and added:
> 
> "add_type_embeddings": true_
> 
> **Other notes**
> 
> Code paths are still hard-coded, as in the original authors’ version. Also need folder structure as suggested in original repo. This is work in progress, but you can check my fork and let me know if you run into any issues. Need microsoft/BiomedVLP-CXR-BERT-specialized from huggingface at code root folder right now. The results of my run in CT_Rate validation set are in the folder rate_res/. Can run compute_metrics.py to get AUC and other performance metrics.

Could you tell me where you download the totalSegmentator mask for the CT-RATE data? Thank you!

**Comment by lavsendahal** [comment]
Its on CT Rate HF. Here's the link 
https://huggingface.co/datasets/ibrahimhamamci/CT-RATE/tree/main/dataset/ts_seg/ts_total

**Comment by jz5426** [comment]
> Its on CT Rate HF. Here's the link https://huggingface.co/datasets/ibrahimhamamci/CT-RATE/tree/main/dataset/ts_seg/ts_total

Thank you so much!

**Comment by laihaoran** [comment]
@lavsendahal  Thanks a lot for sharing the reproducible code. I followed your implementation and successfully reproduced the reported results (CT-RATE: 77.8, RAD-ChestCT: 68.0) — much appreciated!

However, during further checking I found an important issue: the CT-RATE dataset actually contains 18 categories, but the current fVLM implementation only evaluates 16 of them. Specifically, Medical material and Lymphadenopathy are missing in both CT-RATE and RAD-ChestCT evaluations. In the main text, the results are compared directly with CT-CLIP’s 18-class evaluation, which makes the comparison unfair.

I hope the authors can clarify this discrepancy and update the evaluation to ensure a fair and consistent comparison.

**Comment by windygoo** [comment]
> [@lavsendahal](https://github.com/lavsendahal) Thanks a lot for sharing the reproducible code. I followed your implementation and successfully reproduced the reported results (CT-RATE: 77.8, RAD-ChestCT: 68.0) — much appreciated!
> 
> However, during further checking I found an important issue: the CT-RATE dataset actually contains 18 categories, but the current fVLM implementation only evaluates 16 of them. Specifically, Medical material and Lymphadenopathy are missing in both CT-RATE and RAD-ChestCT evaluations. In the main text, the results are compared directly with CT-CLIP’s 18-class evaluation, which makes the comparison unfair.
> 
> I hope the authors can clarify this discrepancy and update the evaluation to ensure a fair and consistent comparison.
1. The Medical material and Lymphadenopathy diseases are ignored during evaluation because they are outside organs (for  Lymphadenopathy ) or not located within a specific organ region (for Medical material), so the organ masks generated by Totalsegmentor do not cover them and they actually do not participate in the training process. 2. The results of CT-CLIP also excluded these two diseases, based on their v1 paper version at the time of our submission.

Thanks very much for your thoughtful and constructive comments! We will update the results in the later version.

**Comment by zhShengY** [comment]
@lavsendahal May I ask if you have made any changes to the learning rate? The results I ran here are always a few points different from the code published by the author for CT-rate data. The following distribution is my training log and the author's training log.
{
    "run": {
        "task": "image_text_pretrain",
        "lr_sched": "linear_warmup_cosine_lr",
        "init_lr": 0.0001,
        "min_lr": 1e-06,
        "warmup_lr": 1e-06,
        "weight_decay": 0.05,
        "max_epoch": 20,
        "batch_size_train": 14,
        "batch_size_eval": 8,
        "num_workers": 16,
        "warmup_steps": 377,
        "log_freq": 1,
        "seed": 42,
        "output_dir": "../multi-modal-results/pretrain_ckpts",
        "amp": false,
        "resume_ckpt_path": null,
        "evaluate": false,
        "train_splits": [
            "train"
        ],
        "device": "cuda",
        "world_size": 4,
        "dist_url": "env://",
        "distributed": true,
        "rank": 0,
        "gpu": 0,
        "dist_backend": "nccl"
    },
    "model": {
        "arch": "blip_pretrain_vit",
        "load_pretrained": false,
        "pretrained": "https://storage.googleapis.com/sfr-vision-language-research/BLIP/models/model_base_capfilt_large.pth",
        "vit_type": "base",
        "vit_grad_ckpt": false,
        "vit_ckpt_layer": 0,
        "image_size": 224,
        "alpha": 0.5,
        "med_config_path": "BiomedVLP-CXR-BERT-specialized/config.json",
        "embed_dim": 256,
        "prompt": "a picture of ",
        "model_type": "fvlp",
        "ckpt_path": null,
        "load_finetuned": false,
        "max_txt_len": 384,
        "max_txt_len_conc": 200,
        "max_txt_len_desc": 200,
        "queue_size": 0
    },
    "preprocess": {
        "vis_processor": {
            "train": {
                "name": "blip_image_train",
                "image_size": 224
            }
        },
        "text_processor": {
            "train": {
                "name": "blip_caption"
            }
        }
    },
    "datasets": {
        "coco_caption": {
            "dataset_card": "dataset_card/coco_caption.md",
            "data_type": "images",
            "build_info": {
                "annotations": {
                    "train": {
                        "storage": "annotations.json"
                    }
                },
                "images": {
                    "storage": "train_resized_images/"
                }
            },
            "vis_processor": {
                "train": {
                    "name": "blip_image_train",
                    "image_size": 224
                }
            },
            "text_processor": {
                "train": {
                    "name": "blip_caption"
                }
            }
        }
    }
}
{"epoch": "0", "train_lr": "0.0000565", "train_loss": "14.1564693", "train_lung_itc": "3.7623620", "train_heart_itc": "3.9720836", "train_esophagus_itc": "3.7368087", "train_aorta_itc": "3.3269345"}
{"epoch": "1", "train_lr": "0.0000994", "train_loss": "13.9774000", "train_lung_itc": "3.6642210", "train_heart_itc": "3.9366231", "train_esophagus_itc": "3.7312112", "train_aorta_itc": "3.2941042"}
{"epoch": "2", "train_lr": "0.0000976", "train_loss": "13.8605501", "train_lung_itc": "3.5711239", "train_heart_itc": "3.9227486", "train_esophagus_itc": "3.7289641", "train_aorta_itc": "3.2798929"}
{"epoch": "3", "train_lr": "0.0000946", "train_loss": "13.8328195", "train_lung_itc": "3.4792857", "train_heart_itc": "3.9142808", "train_esophagus_itc": "3.7297475", "train_aorta_itc": "3.2665299"}
{"epoch": "4", "train_lr": "0.0000905", "train_loss": "13.6799084", "train_lung_itc": "3.3849542", "train_heart_itc": "3.9133828", "train_esophagus_itc": "3.7272523", "train_aorta_itc": "3.2610812"}
{"epoch": "5", "train_lr": "0.0000855", "train_loss": "13.5693323", "train_lung_itc": "3.2842173", "train_heart_itc": "3.9075712", "train_esophagus_itc": "3.7300197", "train_aorta_itc": "3.2525157"}
{"epoch": "6", "train_lr": "0.0000796", "train_loss": "13.4841239", "train_lung_itc": "3.2006719", "train_heart_itc": "3.9017763", "train_esophagus_itc": "3.7243028", "train_aorta_itc": "3.2376712"}
{"epoch": "7", "train_lr": "0.0000730", "train_loss": "13.3560885", "train_lung_itc": "3.1314015", "train_heart_itc": "3.8948806", "train_esophagus_itc": "3.7228938", "train_aorta_itc": "3.2279789"}
{"epoch": "8", "train_lr": "0.0000658", "train_loss": "13.2829245", "train_lung_itc": "3.0515127", "train_heart_itc": "3.8911182", "train_esophagus_itc": "3.7207854", "train_aorta_itc": "3.2252175"}
{"epoch": "9", "train_lr": "0.0000582", "train_loss": "13.1827890", "train_lung_itc": "2.9663508", "train_heart_itc": "3.8834367", "train_esophagus_itc": "3.7138651", "train_aorta_itc": "3.2143148"}
{"epoch": "10", "train_lr": "0.0000505", "train_loss": "13.0863010", "train_lung_itc": "2.8959647", "train_heart_itc": "3.8816263", "train_esophagus_itc": "3.7189282", "train_aorta_itc": "3.2026367"}
{"epoch": "11", "train_lr": "0.0000428", "train_loss": "12.8678036", "train_lung_itc": "2.8003010", "train_heart_itc": "3.8773400", "train_esophagus_itc": "3.7164020", "train_aorta_itc": "3.1985511"}
{"epoch": "12", "train_lr": "0.0000352", "train_loss": "12.9278897", "train_lung_itc": "2.7224114", "train_heart_itc": "3.8726134", "train_esophagus_itc": "3.7129572", "train_aorta_itc": "3.1898034"}
{"epoch": "13", "train_lr": "0.0000280", "train_loss": "12.7473791", "train_lung_itc": "2.6528477", "train_heart_itc": "3.8655410", "train_esophagus_itc": "3.7128059", "train_aorta_itc": "3.1764012"}
{"epoch": "14", "train_lr": "0.0000214", "train_loss": "12.6352415", "train_lung_itc": "2.5796234", "train_heart_itc": "3.8623982", "train_esophagus_itc": "3.7129493", "train_aorta_itc": "3.1697887"}
{"epoch": "15", "train_lr": "0.0000155", "train_loss": "12.5362144", "train_lung_itc": "2.4992789", "train_heart_itc": "3.8580228", "train_esophagus_itc": "3.7086512", "train_aorta_itc": "3.1580295"}
{"epoch": "16", "train_lr": "0.0000105", "train_loss": "12.4683710", "train_lung_itc": "2.4517744", "train_heart_itc": "3.8544642", "train_esophagus_itc": "3.7099971", "train_aorta_itc": "3.1500290"}
{"epoch": "17", "train_lr": "0.0000064", "train_loss": "12.2953777", "train_lung_itc": "2.3956696", "train_heart_itc": "3.8484108", "train_esophagus_itc": "3.7076875", "train_aorta_itc": "3.1408870"}
{"epoch": "18", "train_lr": "0.0000034", "train_loss": "12.4684088", "train_lung_itc": "2.3589650", "train_heart_itc": "3.8490910", "train_esophagus_itc": "3.7047975", "train_aorta_itc": "3.1401170"}
{"epoch": "19", "train_lr": "0.0000016", "train_loss": "12.4257958", "train_lung_itc": "2.3408722", "train_heart_itc": "3.8489503", "train_esophagus_itc": "3.7030554", "train_aorta_itc": "3.1289484"}

{"epoch": "0", "train_lr": "0.0000565", "train_loss": "14.0430965", "train_lung_itc": "3.7474605", "train_heart_itc": "3.9456383", "train_esophagus_itc": "3.8025498", "train_aorta_itc": "3.5283988"}
{"epoch": "1", "train_lr": "0.0000997", "train_loss": "13.6724522", "train_lung_itc": "3.5857035", "train_heart_itc": "3.8866129", "train_esophagus_itc": "3.7953467", "train_aorta_itc": "3.4816083"}
{"epoch": "2", "train_lr": "0.0000989", "train_loss": "13.6511159", "train_lung_itc": "3.4400216", "train_heart_itc": "3.8522488", "train_esophagus_itc": "3.7908969", "train_aorta_itc": "3.4583686"}
{"epoch": "3", "train_lr": "0.0000976", "train_loss": "13.4239775", "train_lung_itc": "3.3113734", "train_heart_itc": "3.8330372", "train_esophagus_itc": "3.7881798", "train_aorta_itc": "3.4589250"}
{"epoch": "4", "train_lr": "0.0000957", "train_loss": "13.1820305", "train_lung_itc": "3.2101521", "train_heart_itc": "3.8201473", "train_esophagus_itc": "3.7893053", "train_aorta_itc": "3.4463434"}
{"epoch": "5", "train_lr": "0.0000934", "train_loss": "13.1585638", "train_lung_itc": "3.1153560", "train_heart_itc": "3.8032264", "train_esophagus_itc": "3.7864681", "train_aorta_itc": "3.4309504"}
{"epoch": "6", "train_lr": "0.0000905", "train_loss": "13.1367099", "train_lung_itc": "3.0697787", "train_heart_itc": "3.7989420", "train_esophagus_itc": "3.7864303", "train_aorta_itc": "3.4325693"}
{"epoch": "7", "train_lr": "0.0000873", "train_loss": "13.0760879", "train_lung_itc": "2.9676784", "train_heart_itc": "3.7840478", "train_esophagus_itc": "3.7830124", "train_aorta_itc": "3.4211196"}
{"epoch": "8", "train_lr": "0.0000836", "train_loss": "12.8837550", "train_lung_itc": "2.8936944", "train_heart_itc": "3.7727783", "train_esophagus_itc": "3.7859698", "train_aorta_itc": "3.4174257"}
{"epoch": "9", "train_lr": "0.0000796", "train_loss": "12.8917007", "train_lung_itc": "2.8281858", "train_heart_itc": "3.7662500", "train_esophagus_itc": "3.7790546", "train_aorta_itc": "3.4146372"}
{"epoch": "10", "train_lr": "0.0000752", "train_loss": "12.6444871", "train_lung_itc": "2.7518032", "train_heart_itc": "3.7524136", "train_esophagus_itc": "3.7806985", "train_aorta_itc": "3.3961716"}
{"epoch": "11", "train_lr": "0.0000706", "train_loss": "12.5644645", "train_lung_itc": "2.6757337", "train_heart_itc": "3.7378871", "train_esophagus_itc": "3.7825271", "train_aorta_itc": "3.3887193"}
{"epoch": "12", "train_lr": "0.0000658", "train_loss": "12.4291960", "train_lung_itc": "2.6004973", "train_heart_itc": "3.7251897", "train_esophagus_itc": "3.7785712", "train_aorta_itc": "3.3706355"}
{"epoch": "13", "train_lr": "0.0000608", "train_loss": "12.4522321", "train_lung_itc": "2.5570721", "train_heart_itc": "3.7165772", "train_esophagus_itc": "3.7749991", "train_aorta_itc": "3.3683354"}
{"epoch": "14", "train_lr": "0.0000557", "train_loss": "12.2217136", "train_lung_itc": "2.4959611", "train_heart_itc": "3.7097519", "train_esophagus_itc": "3.7689600", "train_aorta_itc": "3.3689636"}
{"epoch": "15", "train_lr": "0.0000505", "train_loss": "12.2413548", "train_lung_itc": "2.4057577", "train_heart_itc": "3.6906699", "train_esophagus_itc": "3.7727290", "train_aorta_itc": "3.3538884"}
{"epoch": "16", "train_lr": "0.0000453", "train_loss": "12.0921114", "train_lung_itc": "2.3239336", "train_heart_itc": "3.6773206", "train_esophagus_itc": "3.7678491", "train_aorta_itc": "3.3394512"}
{"epoch": "17", "train_lr": "0.0000402", "train_loss": "12.1366635", "train_lung_itc": "2.2531691", "train_heart_itc": "3.6645907", "train_esophagus_itc": "3.7769576", "train_aorta_itc": "3.3280480"}
{"epoch": "18", "train_lr": "0.0000352", "train_loss": "11.9871331", "train_lung_itc": "2.1769020", "train_heart_itc": "3.6480589", "train_esophagus_itc": "3.7715614", "train_aorta_itc": "3.3203445"}
{"epoch": "19", "train_lr": "0.0000304", "train_loss": "11.8201054", "train_lung_itc": "2.1028232", "train_heart_itc": "3.6354394", "train_esophagus_itc": "3.7651691", "train_aorta_itc": "3.3061251"}
)

**Comment by windygoo** [comment]
> "max_epoch": 20,

change the max_epoch to 30, and the results will be ok.
